File: app/src/main/java/com/aether/client/OverlayService.java
package com.aether.client;

import android.app.Service;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.PixelFormat;
import android.graphics.Typeface;
import android.os.IBinder;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.TextView;

public class OverlayService extends Service {

    private WindowManager windowManager;
    private TextView button;

    @Override
    public void onCreate() {
        super.onCreate();

        windowManager =
                (WindowManager) getSystemService(WINDOW_SERVICE);

        button = new TextView(this);

        button.setText("A");
        button.setTextColor(Color.WHITE);
        button.setTextSize(18);
        button.setGravity(Gravity.CENTER);
        button.setTypeface(Typeface.DEFAULT, Typeface.BOLD);

        button.setBackgroundColor(Color.rgb(35, 100, 220));

        WindowManager.LayoutParams params =
                new WindowManager.LayoutParams(
                        60,
                        60,
                        WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY,
                        WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
                        PixelFormat.TRANSLUCENT);

        params.gravity = Gravity.TOP | Gravity.START;
        params.x = 30;
        params.y = 200;

        windowManager.addView(button, params);

        button.setOnClickListener(v -> {
            Intent intent =
                    new Intent(this, MainActivity.class);

            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

            startActivity(intent);
        });

        button.setOnTouchListener(
                new View.OnTouchListener() {

                    private int startX;
                    private int startY;
                    private float downX;
                    private float downY;

                    @Override
                    public boolean onTouch(
                            View v,
                            MotionEvent event) {

                        switch (event.getAction()) {

                            case MotionEvent.ACTION_DOWN:
                                startX = params.x;
                                startY = params.y;
                                downX = event.getRawX();
                                downY = event.getRawY();
                                return false;

                            case MotionEvent.ACTION_MOVE:
                                params.x =
                                        startX +
                                        (int)(event.getRawX() - downX);

                                params.y =
                                        startY +
                                        (int)(event.getRawY() - downY);

                                windowManager.updateViewLayout(
                                        button,
                                        params);

                                return true;
                        }

                        return false;
                    }
                });
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        if (button != null) {
            windowManager.removeView(button);
        }
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
          }

File: app/src/main/java/com/aether/client/OverlayService.kt
package com.aether.client

import android.app.Service
import android.content.Intent
import android.graphics.Color
import android.graphics.PixelFormat
import android.graphics.Typeface
import android.os.IBinder
import android.provider.Settings
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.view.WindowManager
import android.widget.TextView

class OverlayService : Service() {

    private lateinit var windowManager: WindowManager
    private lateinit var button: TextView

    override fun onCreate() {
        super.onCreate()

        windowManager =
            getSystemService(WINDOW_SERVICE) as WindowManager

        button = TextView(this).apply {
            text = "A"
            textSize = 18f
            setTextColor(Color.WHITE)
            gravity = Gravity.CENTER
            typeface = Typeface.DEFAULT_BOLD
            setBackgroundColor(Color.rgb(135, 95, 255))
        }

        val params = WindowManager.LayoutParams(
            65,
            65,
            WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY,
            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
            PixelFormat.TRANSLUCENT
        )

        params.gravity = Gravity.TOP or Gravity.START
        params.x = 25
        params.y = 250

        windowManager.addView(button, params)

        button.setOnClickListener {
            val intent = Intent(this, MainActivity::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            startActivity(intent)
        }

        button.setOnTouchListener(
            object : View.OnTouchListener {

                private var startX = 0
                private var startY = 0
                private var downX = 0f
                private var downY = 0f

                override fun onTouch(
                    view: View,
                    event: MotionEvent
                ): Boolean {

                    when (event.action) {

                        MotionEvent.ACTION_DOWN -> {
                            startX = params.x
                            startY = params.y
                            downX = event.rawX
                            downY = event.rawY
                            return false
                        }

                        MotionEvent.ACTION_MOVE -> {

                            params.x =
                                startX +
                                (event.rawX - downX).toInt()

                            params.y =
                                startY +
                                (event.rawY - downY).toInt()

                            windowManager.updateViewLayout(
                                button,
                                params
                            )

                            return true
                        }
                    }

                    return false
                }
            }
        )
    }

    override fun onDestroy() {
        if (::button.isInitialized) {
            windowManager.removeView(button)
        }

        super.onDestroy()
    }

    override fun onBind(intent: Intent?): IBinder? {
        return null
    }
}
